### ISSUE TITLE
MSExperiment getMSLevel should not rely on member

### ISSUE THREAD

**Issue by timosachsenberg** [OpenMS/OpenMS#6773]
The current design of MSExperiment allows to filter e.g., all MS1 level. getMSLevel would still return MS1 as one of the contained level because it uses an internal member variable.
This leads to hard to debug bugs.

TODO:
-  determine current MS levels every time the function is called! and document that this is potentially a costly operation

e.g.
```
set<UInt> level;
for (const auto& s : spectra) { s.insert(s.getMSLevel(); }
return std::vector(level.begin(), level.end());
```

**Comment by timosachsenberg** [comment]
see https://github.com/OpenMS/OpenMS/pull/6774

**Comment by timosachsenberg** [comment]
Seems it is only set if updateRanges is called.
a minimal solution that prevents "unitialized" use would be to throw an error if empty.
But this doesn't solve the problem that any other filtering will invalidate this member.
Maybe something we should discuss in the future.
For now I will at least add the exception

**Comment by cbielow** [comment]
for MS-level you can certainly get away with computing it on-the-fly when calling `getMSLevels()`. No harm done.
However, all the other ranges are somewhat expensive and should not be computed on the fly IMHO.

For consistency we could:
1) when loading an MSExperiment from file, every `.load()` ensures that ranges (incl. MS level) are properly updated.
2) when the user modifies an MSExperiment (or builds it from scratch), he needs to call `updateRanges()`

This seems like a reasonable tradeoff to me.
For consistency, I'd include MS-level in the above, especially because code like:
`for (auto l : exp.getMSLevels())` would potentially break, since you get a temporary and not a reference (depending on the underlying implementation of range-for), if I remember the discussion on that topic correctly.

**Comment by timosachsenberg** [comment]
> for MS-level you can certainly get away with computing it on-the-fly when calling `getMSLevels()`. No harm done (well, see below)/ However, all the other ranges are somewhat expensive and should not be computed on the fly IMHO.

Wouldn't returning a MSExperimentRanges be a better design then and more transparent?
E.g., creating these on demand (or e.g. during loading).

> For consistency we could:
> 1. when loading an MSExperiment from file, every `.load()` ensures that ranges (incl. MS level) are properly updated.
I think this is really necessary. At least we could do that. But see my suggestion above. I think this is a cleaner approach.

> 2. when the user modifies an MSExperiment (or builds it from scratch), he needs to call `updateRanges()`
> This seems like a reasonable tradeoff to me. For consistency, I'd include MS-level in the above, especially because code like: `for (auto l : exp.getMSLevels())` would potentially break, since you get a temporary and not a reference (depending on the underlying implementation of range-for), if I remember the discussion on that topic correctly.

One level of temporary is fine.

**Comment by cbielow** [comment]
> Wouldn't returning a MSExperimentRanges be a better design then and more transparent?
E.g., creating these on demand (or e.g. during loading) and returning where created.

And then you have to pass down two objects into whatever function needs an MSExperiment + ranges?
The user still has to ensure that the range object is up to date, but now he also has to pass the correct one...
I'd say this leaves even more room for error.
You can argue that users can create that object on the fly, if they need it, but this may have really bad implications for speed if used wrongly and if you want speed you need to persist the range-object, and then you back to square one (how do you know it still matches the MSExp?).

**Comment by jpfeuffer** [comment]
You can also add a param to the constructor of MSExperiment  "auto_update_ranges".
If so, it will be done for every filtering. Of course, you would need to restrict direct access to internals and only allow modification via methods of MSExperiment (but this should be a better design anyways if this is not the case yet).

**Comment by timosachsenberg** [comment]
> > Wouldn't returning a MSExperimentRanges be a better design then and more transparent?
> > E.g., creating these on demand (or e.g. during loading) and returning where created.
> 
> And then you have to pass down two objects into whatever function needs an MSExperiment + ranges? The user still has to ensure that the range object is up to date, but now he also has to pass the correct one... I'd say this leaves even more room for error. 

Fair point.  For immutable objects, I would agree completely. For mutable ones, it seems we can't have both performance and safer code. Other, more complex designs that try to enforce encapsulation are also not very appealing.

> You can argue that users can create that object on the fly, if they need it, but this may have really bad implications for speed if used wrongly and if you want speed you need to persist the range-object, and then you back to square one (how do you know it still matches the MSExp?).

yeah I see the point. It just seems a bit bad that we allow the object to be in an "undefined/invalid" state.

**Comment by timosachsenberg** [comment]
another thing:
two spectra do not compare equal depending on whether updateRanges was called or not.
Fell into this trap once...

**Comment by jpfeuffer** [comment]
I say the same subclassing mechanism should exist for MSSpectrum. There should be a SortedMSSpectrum and an AutoRangeMSSpectrum etc. Because then you can enforce specific Preconditions and users can choose which type they need.

**Comment by cbielow** [comment]
>two spectra do not compare equal depending on whether updateRanges was called or not.

Simply a bug that needs fixing.

> I say the same subclassing mechanism should exist for MSSpectrum. There should be a SortedMSSpectrum and an AutoRangeMSSpectrum etc (where sorted is of course even stricter than AutoRange). Because then you can enforce specific Preconditions and users can choose which type they need.

the implications of this would be very hard to enforce throughout the library.
Think about all the polymorphism / conversions that you need to implement to make this work smoothly. I don't see anyone taking the time to implement this. And even if, I'm not sure that we reduce the total number of errors (slicing, other conversion errors, breaking APIs everywhere, requiring way more templates, increased compile time, ... etc etc)
This needs way more thought IMHO.
On first sight: The default should be a sorted, "ranged" MSExp. This is what you can expect when being handed an MSExperiment, unless its very clear from context that this is not the case. Every algorithm messing with these properties should recreate these invariants if handing on the MSExperiment (e.g. a PeakPicker, NoiseFilter etc). No other magic needed.

**Comment by jpfeuffer** [comment]
> Simply a bug that needs fixing.
Agree

> the implications of this would be very hard to enforce throughout the library.
IMHO easier than enforcing that everyone updatesRanges all the time.

> Think about all the polymorphism / conversions that you need to implement to make this work smoothly. I don't see anyone taking the time to implement this.
Yes would be a bigger undertaking.

> breaking APIs everywhere
As always

> requiring way more templates, increased compile time
Not necessarily if you work on the baseclass. They all might actually have the same members and functions, just different implementations of the functions that enforce specific preconditions, so maybe not even slicing.

> On first sight: The default should be a sorted, "ranged" MSExp.
I agree and would be fine with just having that. But we need to go through the codebase once and ensure that this is the case and take special care when reviewing PRs (or find an automated way to check that: compile flag in nightlies, with certain locations ignored?)

**Comment by timosachsenberg** [comment]
Fixed. Are calculated on the fly
